### DISCUSSION TITLE
How can I get better png compression (lower file size)

### DISCUSSION THREAD

**Discussion by martonx** [SixLabors/ImageSharp#1599] [Q&A] [answered]
Hi,

I have a lot of big png images. When I compress them with tinypng, I got a bit smaller images then compress them with ImageSharp.

This is my PngEncoder:

        var encoder = new PngEncoder()
        {
            CompressionLevel = PngCompressionLevel.DefaultCompression,
            TransparentColorMode = PngTransparentColorMode.Clear,
            BitDepth = PngBitDepth.Bit8,
            ColorType = PngColorType.Palette
        };

Could you advice anything else, what could I set in options, to get more 10-20% lesser file sizes?
I tried PngCompressionLevel.BestCompression but it caused only a marginally smaller file.

**Comment by JimBobSquarePants** [comment, marked as answer]
I'd have to analyze both the input and output from tiny png to compare to give you accurate instruction. If you could post some examples that would be really great. 

For starters though I would change the filter method to None as that's better for palette based images, also bump that compression.

Compression tools tend to remove metadata so I would check that and delete to match also.

**Comment by martonx** [comment]
Thank you for your answer.

Here it is one original file:
![test_original](https://user-images.githubusercontent.com/407433/114979674-9907cc80-9e8b-11eb-9324-c8e9e15682fb.png)

This is the tinypng result (142 Kbyte):
![test_original](https://user-images.githubusercontent.com/407433/114979711-a6bd5200-9e8b-11eb-85d3-84498b12edaf.png)

This is my ImageSharp result (188 Kbyte):
![test_resized](https://user-images.githubusercontent.com/407433/114979773-c2c0f380-9e8b-11eb-8174-0d466bb8db13.png)

My encoder config is currently:
      CompressionLevel = PngCompressionLevel.BestCompression,
      TransparentColorMode = PngTransparentColorMode.Clear,
      BitDepth = PngBitDepth.Bit8,
      ColorType = PngColorType.Palette,
      FilterMethod = PngFilterMethod.None,
      IgnoreMetadata = true

**Reply by JimBobSquarePants** [reply]
Thanks. I had a look and the best we can do currently is the output you've created we might be able to improve a little once we merge the new zlib compression stream I'm working on but the main difference is actually due to the quality of output. If you do a side by side comparison you'll see that the TinyPng version is washed out, losing a lot of color information. This means there's larger areas of the same pixel data which leads to better compression.

You can see it clearly in the top/left when we align the images like below. Lots of the gradient is simply gone.

Our output does not make that sacrifice (_because people would definitely complain if we did!_) so it's a much, much better match.

You would have to use your own quantizer (_perhaps a modified median cut_) to more closely match their output and allow compression to the same level.

**TinyPng (Left) - ImageSharp (Right)**
![image](https://user-images.githubusercontent.com/385879/115089003-4cfb6d00-9f09-11eb-9a01-3b2695375b0e.png)
**ImageSharp (Left) - Original (Right)**
![image](https://user-images.githubusercontent.com/385879/115089198-a6639c00-9f09-11eb-8a2f-22dc0b431a18.png)

**Reply by martonx** [reply]
Thank you @JimBobSquarePants ! Ok, I can live with current image compression result. Do you know any ETA for this new zlib compressor?

**Reply by JimBobSquarePants** [reply]
No eta I'm afraid. It's a complicated space and I have a lot of testing to do.

**Reply by saucecontrol** [reply]
Just thought I'd point out, the TinyPng output here is correct.  The color difference you're seeing isn't due to poor quantization but rather due to the fact the original image has an Apple (gamma 1.8) color profile.  TinyPng is converting the image to sRGB and dropping the profile, while ImageSharp  (with `IgnoreMetadata = true`) is just dropping the profile, causing it to be incorrectly interpreted as sRGB.  Likewise, Beyond Compare isn't color aware, so it shows the images without gamma adjustment.  If you look at the embedded images in the top comment more closely, you should see that the original and TinyPng versions match while the ImageSharp one is darker (at least they show up that way in Chrome).

**Reply by JimBobSquarePants** [reply]
Didn’t know that about Beyond Compare! There’s additional massive data loss though to my eye. Areas that are flat that shouldn’t be.

I’d love to get our color handling improved but will need someone else to complete the open PR.

**Reply by saucecontrol** [reply]
Yeah, that's a quirk of the human visual system. Our eyes really like contrast, so when you show someone an image with boosted contrast (a deeper gamma curve will do that), they'll almost always think the original looks overly flat/dull by comparison.  Here's the original converted to sRGB using the Adobe CMS and then saved as a 32-bit PNG:
![applergbthing_srgb](https://user-images.githubusercontent.com/7929102/119236583-093de980-baed-11eb-9642-4bb8c47ec77f.png)

Without even looking, I'd guess TinyPng is using pngquant, which does an amazing job with low-color gradients (its weakness is high-color images).

I had a look at the open color conversion PR, but it's a monster.  I need to figure out LUT-based profile conversion myself, so I'll give it a closer look once I've got that worked out on my end.

**Reply by JimBobSquarePants** [reply]
Interesting…. Yeah I open that PR once in a while then almost immediately close it. Far too big for me to take on just now!

**Comment by antonfirsov** [comment]
@martonx just a sidenote that even if you will be able to find optimal parameters for this particular PNG, there's no guarantee they will be perfect for another one. Tools like tinypng or optipng work on trial and error basis, encoding the image with different parameters until they find the best parameter set for a given PNG.

Although you will be probably able to find a good-enough solution with palette encoding, you probably need to accept that there will be always some difference in comparison to those tools if you need a solution that works with an arbitrary PNG.
